### ISSUE TITLE
Support generating non-source files

### ISSUE THREAD

**Issue by geeknoid** [dotnet/roslyn#57608]
We have a few use cases where we'd like to leverage source generators to produce non-source files. Basically, we want to use source generators to generate secondary build outputs. For example:

* We annotate data models with data classification attributes to denote PII and we produce a CSV file that enumerates the types and members in the project and their data classification. You can then take this CSV file and use it when doing privacy audits.

* We use source generators to produce a bunch of code around metrics. We'd like to emit a schema file describing the effective shape of the metrics being produced by the code. This would probably be in the form of an output JSON file.

We have a few other scenarios in the wings which could leverage this approach.

Unfortunately, source generation is currently limited to C# sources. Trying to emit files with other extensions doesn't work. As a result, our source generator currently contains this horrible hack:

```csharp
        public void Execute(GeneratorExecutionContext context)
        {
            ... generate the CSV file ...

            context.CancellationToken.ThrowIfCancellationRequested();

            /// Adding the workaround to generate reports via File writes since <see cref="GeneratorExecutionContext.AddSource(string, CodeAnalysis.Text.SourceText)"/>
            /// has an underlying check for `.cs` files and automatically adds the `.cs` suffix if the provided filename has some other filetype.
            /// Refer <see href="https://github.com/dotnet/roslyn/blob/v3.8.0/src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs#L63">AppendExtensionIfRequired</see>.
            _directory ??= FileUtilHelpers.GetOutputDirectoryForGeneratedFile(typeof(Generator), context.Compilation.Assembly, context.Compilation.Options.OptimizationLevel);
            _ = Directory.CreateDirectory(_directory);

            // Write properties to CSV file.
            File.WriteAllText(Path.Combine(_directory, _propertiesFileName), properties);

            // Write log methods to CSV file.
            File.WriteAllText(Path.Combine(_directory, _logMethodsFileName), logMethods);
      }
 ```

Could we get 1st class support for producing more than .cs files?

**Comment by ryzngard** [comment]
Are these outputs also included in the final compilation somehow, or are they just independent artifacts of the build?

**Comment by geeknoid** [comment]
They are secondary outputs of the build process, just like the .xml doc file produced for XML doc comments. If C# didn't support XML doc comments, you could in fact implement them fully as a source generator.

Ideally, the output files would be produced in the same directory as the final output assembly, rather than being buried somewhere in the obj directory. But I can live with the obj directory output.

**Comment by lsoft** [comment]
probably, something related? https://github.com/dotnet/roslyn/issues/52677

**Comment by vpenades** [comment]
I agree with this proposal. in my case I would need something as simple as this:

```c#
void AddContent(string hintName, byte[] content); // add file as a content item
```

In the grand scale of things, I would like to have is a framework similar to source generator, but for content files... basically a "content generator". I am not sure if this can be achieved by simply adding an "AddContent" method, or it would require a completely separated, and more specialised framework.

**Comment by AaronRobinsonMSFT** [comment]
This is also of interest to the Interop team and the [DllImport source generator](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Runtime.InteropServices/gen/DllImportGenerator).

We would like to emit artifacts that could be used in other languages and likely used in a post-compile MSBuild Target. Roslyn could give us a `Stream` instance to use or we could call an API that takes a `Stream`. Either approach works for the data stream. The name of the output artifact file would also be important. The specific path isn't interesting coming out of the compiler, since we would move/copy it to another location, but the name would be helpful.

From within MSBuild, it would be helpful to either have a convention of a new set of `Item`s (e.g., `<GeneratorName>Artifacts`) that we could then consume. I would caution against requesting an `Item` name from the generator for the typical name conflict issues. 

/cc @dotnet/interop-contrib @chsienki @jaredpar @kg

**Comment by chsienki** [comment]
@AaronRobinsonMSFT This is basically the design we landed on for the feature :) 

We actually have a (somewhat out of date) PR to implement this on the analyzer side ttps://github.com/dotnet/roslyn/pull/49046  but it's been stuck in limbo due to lack of prioritization. 

Feel free to bug @jaredpar if this is something you think is important to the success of the DllImport generator.

**Comment by stephentoub** [comment]
As a +1, I'd also like this for the RegexGenerator.  I'd like to be able to output non-compiled files that a developer could browse to in solution explorer to be able to provide additional diagnostic information about the regular expressions being used.  Imagine, for example, if in addition to outputting RegexGenerator.g.cs that included the source code, it could also output a .dgml file containing a graph of the DFA for the regex, and a developer could simply double-click that .dgml file to have it opened in VS to visualize the shape of their regex.  Today the best I can do here seems to be emitting that dgml content as a comment embedded in the .cs file file, at which point a developer needs to find it, copy/paste it to a separate file, and then open that file they created.

**Comment by geeknoid** [comment]
BTW, if you think about it, the 20 year old support for XML doc comments in C# could now be totally implemented as one of these plugin "code" generators. I think that helps create a good mental model for what this kind of generic "produce some output files" functionality can offer.

**Comment by ChrML** [comment]
This would also be really useful for projects like Swagger. It could have a content generator that would output OpenAPI data for the API controllers in the project at compile-time rather than runtime. Thus avoiding reflecting and parsing the XML file at runtime.

Even better if the API let you choose if the content should be added as a file next to the build output, or if it should be embedded as a part of the assembly dll file.

**Comment by jaredpar** [comment]
@ChrML in the Swagger case how would you use the output though:

- Have an MSBuild task that copied it somewhere? 
- Tell users where it is on disk and let them consume it at will? 
- etc ...

**Comment by ChrML** [comment]
> @ChrML in the Swagger case how would you use the output though:
> 
> * Have an MSBuild task that copied it somewhere?
> * Tell users where it is on disk and let them consume it at will?
> * etc ...
The exact same way as any solution item marked as "Content". Except in this case the content is not a solution item, but dynamically provided during compile by the source generator.

As for the Swagger-like use-case there could be no change to the code for the library consumers. Internally all the runtime code could be almost-zero, and the library would just serve the compile-time generated OpenAPI file directly to the API.

Assuming ofcourse that there's no dynamic logic involved and all metadata is available at compile-time. I would think this is the most common scenario.

**Comment by BertanAygun** [comment]
Adding my +1 as well for VisualStudio.Extensibility, we would love to have this natively supported since we generate both source files and also a metadata asset to be included as part of the extension build output. So being able to call a method like AddContent(string relativePath, <content>) which handles copying the file to the correct output location like a Content item would be great.

**Comment by kg** [comment]
Things like shader compilers also need to typically generate manifests or metadata in addition to shader binaries - I can imagine someone wanting to do a thing where you write shaders in C# or in string literals, and then a source generator would produce C# methods that expose access to the shader and then embed shader binaries and metadata for use at runtime. The alternative is manually pulling out manifest resources etc, which is error prone (especially since embedding manifest resources exposes you to msbuild's broken dependency tracking).

**Comment by MortenChristiansen** [comment]
I would love this for generating SQL change scripts based on classes representing tables.

**Comment by CollinAlpert** [comment]
I'd also like this to generate Razor pages with Source Generators. Since it's probably too late for .NET 7, can anyone say if this is triaged for .NET 8?

**Comment by andrew-hampton** [comment]
Adding a +1 to this proposal. We have a few scenarios where we launch a separate executable, use reflection to generate a content file and then a separate target to include them. This process works, but can be fragile. Would be fantastic to be able to utilize the context within source generators to achieve this.

**Comment by BlackGad** [comment]
yeah additional scenarios that is not covered:
- embedded resources
- content files
- special files with appropriate compile option
- simple output text files.

The old t4 generation still more functional :(

**Comment by DomenPigeon** [comment]
+1 In my case I would like to utilize the SG for generating JavaScript files, as we have some mapping between the C# backend classes and the frontend which is done in React.

**Comment by alirezapooneh** [comment]
> +1 In my case I would like to utilize the SG for generating JavaScript files, as we have some mapping between the C# backend classes and the frontend which is done in React.

We use **Typewriter** extension to create ts files which is generated from view models and also api controllers as angular service.
But generating by SG is very flexible and better than use other extensions.

**Comment by BlackGad** [comment]
Currently it is possible to write any text into cs commented file. Assign it unique extension and intercept on project. Read content, remove comment markers and save/embed as new item. But still it is bypass.

**Comment by vpenades** [comment]
> Currently it is possible to write any text into cs commented file. Assign it unique extension and intercept on project. Read content, remove comment markers and save/embed as new item. But still it is bypass.

regardless of the text or the extension, the file is added to the project as a Compile item, so msbuild will still handle the file as a compilable .cs file.

what is being requested [here](https://github.com/dotnet/roslyn/issues/57608#issuecomment-1465227377) is not only changing the content or the extension, but also the way msbuild handles the file:  for example as Content, or as  Embedded Resource.

Also, I am curious if it could be possible to generate WPF/Maui XAML files and if the compiler would handle them correctly

**Comment by BlackGad** [comment]
Generally you are right. One thing - code generation optimized for VS code completions and analysis plus some pure integration into IDE. Don't know why they can't handle such simple feature for so long time. I mean cache generators on IDE level when you have internal project with generators is genius) you need to restart IDE to have updated references.

**Comment by BertanAygun** [comment]
@MeikTranel, while I agree that the capabilities exist, in our scenarios we create both code files and content files from the same compilation. 

So for our code generation process, we would have to invoke Roslyn compilation twice to get the syntax and semantic tree making build twice longer and it would also require maintaining 2 different assemblies doing fairly similar things, one for MSBuild, one for Roslyn. It would also mean we have to ensure the msbuild tasks need to be tested across different types of projects which is an added complication.

Things get even slower if there are multiple generators running on the build as MSBuild tasks since due to their isolated nature, they would all have to invoke Roslyn individually.

**Comment by mhmd-azeez** [comment]
I'd also like to add this use case:

I am trying to write a C# PDK for [http://extism.org](https://t.co/IysQVgHBTW). So far I have been able to implement a very basic plugin: https://github.com/extism/csharp-pdk.

The problem is, Extism plugins need to export functions, and the .NET WASI SDK doesn't allow that yet (I think)...

So for every function (e.g. https://github.com/extism/csharp-pdk/blob/master/samples/SamplePlugin/Functions.cs#L9)

I need to write the native glue code (e.g. https://github.com/extism/csharp-pdk/blob/master/samples/SamplePlugin/native.c#L10).

Now I want to know if it's possible to generate the glue code automatically using source generators...
But the glue code has to be referenced in the plugin csproj: https://github.com/extism/csharp-pdk/blob/master/samples/SamplePlugin/SamplePlugin.csproj#L20

**Comment by CyrusNajmabadi** [comment]
Could you just write a tool that runs post-compile and produces teh 'c' code?  It seems like using SGs here would be undesirable as they would run (potentially expensively), only to produce nothing that would actually impact the compilation.

**Comment by mhmd-azeez** [comment]
@CyrusNajmabadi that could work, but I want the user to be able to specify which functions they want to export (e.g. via an attribute). Source Generators make really this easy because they give you access to the `Compilation`. Without them I'd have to somehow parse all of the C# files to look for these exported functions

**Comment by CyrusNajmabadi** [comment]
the code will be compiled at that point.  You can either trivially use reflection to find your attributes, or you can just use roslyn directly again from your tool.  

This has the added benefit of only running during builds, and not ever at all in the IDE.

**Comment by mhmd-azeez** [comment]
Interesting, is there any similar tool I could take a look on github? this whole area is new for me so I'd appreciate any help I can get

**Comment by mhmd-azeez** [comment]
@CyrusNajmabadi thank you very much, I'll try that route too

**Comment by AaronRobinsonMSFT** [comment]
> You'd just need to write a simple console app. You'd invoke that from your build, passing in the result dll of the prior step. You'd be able to use reflection to read it using normal reflection APIs in .net.

This is what DNNE does - https://github.com/AaronRobinsonMSFT/DNNE. I read in raw metadata but the idea is the same.

**Comment by mhmd-azeez** [comment]
@AaronRobinsonMSFT thank you very much! this is very helpful

**Comment by MeikTranel** [comment]
Two more use cases that were entirely covered by what MSBuild already giveths.

I can throw mine in for good measure but the task is basically a .NET Framework version of @AaronRobinsonMSFT's DNNE:
https://github.com/MeikTranel/NXPorts/tree/master/src/NXPorts

@BertanAygun I'd have to actually look at the two problems to qualify if accessing the roslyn workspace twice would be to much of a burden, but i can see your point.

@BlackGad Tuning tasks to be IDE aware is also a solved problem in the task ecosystem - the IDE/Designer can and DOES already run tasks to generate stuff at designtime - you do not have to manually build or restart VS for anything if you do it right. The whole pre/postcompile .bat script dogma is what used to incur this problem in the first place.

**Comment by BlackGad** [comment]
@MeikTranel nope. VS UI for assemblies that relay on your source generator assembly will not accept changes until you reload while IDE it. This is for analyzer issues list, generated files list and code completion in editor.

**Comment by dotMorten** [comment]
Here's another scenario: I have a symbol font, and I generate a C# enum and helper classes to easily access those characters, but I also want to generate a XAML ResourceDictionary to easily be able to pull in icons as static resources.

**Comment by dotMorten** [comment]
@CyrusNajmabadi You'd be surprised. Fonts are just one. We have a long range of styles and iconography resources that changes and evolves almost weekly. I did write a prebuild-commandline tool that does it, but source generators is a much better way to do this scenario.

**Comment by dotMorten** [comment]
That’s problematic. It means duplicating resources and not just have one source of truth. It means remembering to run the tool when your submodules update. I use plenty of Microsoft made code generators where what they use to generate on rarely changes and we aren’t questioning those. Why is my scenario different?

**Comment by CyrusNajmabadi** [comment]
> That’s problematic. It means duplicating resources and not just have one source of truth. 

I would just have your single tool be the source of truth.

> It means remembering to run the tool when your submodules update. 

I don't think you'd have to remember, it would just fallout of having a build system.  The build system would see the inputs change, and would rerun your tool.

> I use plenty of Microsoft made code generators where what they use to generate on rarely changes and we aren’t questioning those.

I am definitely questioning those *regularly*.  I don't want to compound bad decisions with more badness :)

> Why is my scenario different? 

Because it's definitely doing things that indicate that it doesn't fall under an appropriate scenario for SGs.  That tehre may be other existing htings that likely shouldn't have been SGs either doesn't mean the philosophy should be "we should keep expanding", it means "we should be better about getting existing things to use better options". :)

**Comment by BlackGad** [comment]
> > That’s problematic. It means duplicating resources and not just have one source of truth. 
> 
> I would just have your single tool be the source of truth.
> 
> > It means remembering to run the tool when your submodules update. 
> 
> I don't think you'd have to remember, it would just fallout of having a build system.  The build system would see the inputs change, and would rerun your tool.
> 
> > I use plenty of Microsoft made code generators where what they use to generate on rarely changes and we aren’t questioning those.
> 
> I am definitely questioning those *regularly*.  I don't want to compound bad decisions with more badness :)
> 
> > Why is my scenario different? 
> 
> Because it's definitely doing things that indicate that it doesn't fall under an appropriate scenario for SGs.  That tehre may be other existing htings that likely shouldn't have been SGs either doesn't mean the philosophy should be "we should keep expanding", it means "we should be better about getting existing things to use better options". :)

So why do you need source code generator? Write a tool which will do this for you on demand. How it is differs? SG generates only when something changed in the code, use tool! Generate partial class using Roslyn analysis is not completed thing. But you are here, right?

**Comment by CyrusNajmabadi** [comment]
> So why do you need source code generator? 

Who are you referring to here?

**Comment by BlackGad** [comment]
Xaml code is the same part of codebase for your app, same for embedded resources, content files, additional resource files that must be generated to output folder. Why static generated Json serialiser code is differs from the open API scheme? So, please, do not project your vision too all the cases. SG cannot generate some code from which application consists of.

**Comment by CyrusNajmabadi** [comment]
The reason to use SGs is because the generated code both needs to analyze what the *user* has provided in their source, *and* then use that to generate code that is then compiled *with* that user code.  It exists as there's fundamentally an unresolvable circularity otherwise.  Namely that if you tried to write an external tool, you wouldn't be able to analyze things properly (since the user code would be incomplete), and then you wouldn't have a way to inject results back in to make it complete.

That doesn't apply to generating non-source files.  The user code cannot semantically depends on such things (there's no such concept in our language).  So, using this system to accomplish that goal is both unnecessary, and practically speaking, the most expensive way to accomplish this task.  So it's explicitly not supported so that tools that need to do this properly remove themselves from thsi part of the pipeline and insert themselves at better parts, without such costs.

**Comment by BlackGad** [comment]
> The reason to use SGs is because the generated code both needs to analyze what the *user* has provided in their source, *and* then use that to generate code that is then compiled *with* that user code.  It exists as there's fundamentally an unresolvable circularity otherwise.  Namely that if you tried to write an external tool, you wouldn't be able to analyze things properly (since the user code would be incomplete), and then you wouldn't have a way to inject results back in to make it complete.

That's the point! Proper and convinion analysis of the source code, based on which you able to generate xaml that then could be used to generate back c# files. Or generate embedded resources with proper names based on C# class/enum values, attribute values.

**Comment by CyrusNajmabadi** [comment]
> That's the point! Proper and convinion analysis of the source code,

That's not what the post i was responding to said it was for.  It mentioned it was for analyzing a symbol font.  I don't see any circularity there. As such, my position is that this space is better handled as a pre or post build step.

**Comment by dotMorten** [comment]
> the build system would see the inputs change, and would rerun your tool.

Errr that's what a code generator does. I fail to see the difference here. Seems a lot cleaner to just use the source generators than building a custom tool that does the exact same thing.

> The reason to use SGs is because the generated code both needs to analyze what the user has provided in their source, and then use that to generate code that is then compiled with that user code

If that's the case I don't get why we got the `AdditionalFiles`  feature. In my case I'm generating XAML. XAML _is_ code. Whether I generate C# or XAML shouldn't really matter, and there could be a wealth of various inputs that means I want to generate XAML - even cases that fits your more limiting scope. I could for instance be auto-generating datatemplates based on C# view models.

**Comment by CyrusNajmabadi** [comment]
> Errr that's what a code generator does. I fail to see the difference here. Seems a lot cleaner to just use the source generators than building a custom tool that does the exact same thing.

Source generators come with enormous costs of their own, including those that then impact compiling code, as well as all features within an IDE.  Those costs are ok to pay when there is no alternative, and hte circularity issue is inherent to the problem domain.  But those costs are superfluous and unpleasant when not necessary.

> If that's the case I don't get why we got the AdditionalFiles feature.

Because those serve as an input that the compiler itself needs.  It's how things like options flow into analyzers/generators.

As before though, that there is a narrow use case for this is not a reason for anything to use this (or expand to different use cases that don't need circularity support).

**Comment by dotMorten** [comment]
> Because those serve as an input that the compiler itself needs

All the examples and doc I've seen has been to provide the input to the source generator, not the compiler. The CSWin32 generator is a good example of that. In my case I have a list of glyphs for which I want resources and enums generated.

> Source generators come with enormous costs of their own

Compared to compiling a tool and running it as a pre-build set on each build at the time time in the build process, and asking people to create all the infrastructure to set this up in msbuild, handle datestamps to ensure things only gets regenerated when needed etc just seems silly when we have a tool that pretty much does it all except this one thing missing: Setting the build action on the generated output.

I don't think it is an unreasonable ask to be able to generate C# and XAML resources that works in conjunction with each other (the symbol-font scenario was just one example - I've mentioned others above)

**Comment by Khitiara** [comment]
generating both at the same time without duplicating the analysis part, as well as the comparative ease of making a source generator by comparison to making a proper build-time tool and configuring msbuild etc to actually run that is the main appeal here i think. if it was as easy to make a post-build or pre-build tool and wire that into the build system as it is to make a quick source generator this issue wouldnt be as big a deal

**Comment by CyrusNajmabadi** [comment]
>  as the comparative ease of making a source generator by comparison to making a proper build-time tool and configuring msbuild etc to actually run that

I mean, that's solvable with a proper investment in "making a proper build time tool and configuring msbuild" easy.  That another ill-fitted system is easy is not a justification to jam more into it when it's not the right fit for the problem space.   

FWIW, i've done pre/post-build stuff before, and it was very easy.  You just define inputs/outputs, and the command line to invoke and it just seems to work fine.  Maybe some templates could help out here?

**Comment by CyrusNajmabadi** [comment]
> All the examples and doc

I'm not in charge of that.  I can't help if other people do things and write docs for things i think are not good usages of hte system.  In other words, i'm not swayed by people using the system beyond the boundaries of what it is intended for.  And, if that happens, and limitations are reached, that's not a motivation to break those down *esp* given how gargantually complex and performance critical (And detrimental) these systems already are.

**Comment by CyrusNajmabadi** [comment]
> Compared to compiling a tool and running it as a pre-build set on each build at the right time in the build process, and asking people to create all the infrastructure to set this up in msbuild,

I don't see how that's more complex than asking *us* to do precisely that, just within Roslyn itself.  You're getting a team to own this, and effectively saying "you go figure out how to make it awesome and solve all the perf issues/etc." :)  I'm pointing out that with the *existing* system we're already having huge issues, and that's *with* trying to keep things limited in terms of hte support set *and* trying to get people to scale back with the unnecessary usages of this system.  

> handle datestamps to ensure things only gets regenerated when needed 

I don't know what this means.  In my experience this just comes for free.  You specify inputs/outputs, and the build system is the one that does the timestamp/up-to-date checking.

> etc just seems silly when we have a tool that pretty much does it all except this one thing missing: Setting the build action on the generated output.

It's not silly if everything is free.  But it's def not free.  The complexity here is already massive.  The perf and scaling issues are pretty much the #1 (or #2) issue impacting us currently.  Throwing more things in *especailly* when they don't need the *core* capability that SGs are intended to provide, is great for people who don't have to worry about these problems, but it's *really* bad for those that do.  

We need to be scaling *way* back in this space, so that what we provide we can provide efficiently and effectively (like what we do with analyzers).  And if that means investments in *other* tools to make them as easy, then that needs to be done, instead of kitchen-sinking SGs.

**Comment by dotMorten** [comment]
>  i've done pre/post-build stuff before, and it was very easy

It's easy for you and me, but we both probably live and breathe msbuild. The system you are talking about is actually what I have right now - It sucks and I changed it over to code generators. It's SO much cleaner and builds faster, but I lost the ability to add XAML. Hence why I came here and added another scenario on here.

I'm not really following your argument about "what the system is intended for". I'm generating code, since XAML _is_ code. The system "boundary" is artificial and whatever we define it to be. If a tool is really great for a job, who cares that it now solves a scenario that wasn't initially in the scope.
There are so many cases where C# code you have goes hand in hand with various UI components, whether that be HTML, XAML or something else used to visualize that data. They are two sides of the same coin. If I throw an attribute on a property, it might want to generate something specific in my type-safe data-template.

> It's not silly if everything is free.

Resource constraints is a valid argument for de-prioritizing. I get that, but you didn't argue from the point of resources. You argued against the scenarios people have presented as being invalid, including the doc and generators Microsoft have put out. T

**Comment by CyrusNajmabadi** [comment]
> I'm not really following your argument about "what the system is intended for"

The system is not intended to be an arbitrary code generator.  It's intended to solve a circularity problem in C#-only codegen scenarios where the generated code needs to know the users non-compiling code in order to figure out what to generate, and then the results of generation are used with the non-compiling code to make compilable code.  That's it. :)

**Comment by CyrusNajmabadi** [comment]
> Resource constraints is a valid argument for de-prioritizing. I get that, but you didn't argue from the point of resources.

I have consistently argued this.  SGs are extremely expensive.  The run in the core level of hte compiler, and are in a dependent path on all compilation/IDE features.  Any costs associated with them are borne by all users at all layers above that stack.  It is not a place to keep adding more costs.  It's a place to reduce costs *literally* as much as possible and leave only for incredibly important scenarios that literally no other approaches would work for.

**Comment by CyrusNajmabadi** [comment]
> but we both probably live and breathe msbuild. 

I consider myself an msbuild novice.  But this domain fits within that novice understanding.  It's just defining a tool (a literal console app) that is given some inputs, options and produces some outputs.  It's basically one of hte oldest concepts possible in programming (and is how things have worked in all other programming domains that have build systems).  For example, it's akin to what you would do in a makefile if you were doing C (or anything else the makefile was controlling).

SGs are about injecting code *into* the compiler *for* the compilation that is happening *right then*.  They are necessarily a plugin system, and not an external tool, because they need to both read *and* write into the very same step that is executing from the build system's perspective.  If your problem domain requires that, then SGs are the answer.  For anything else, i would recommend not using SGs *at all*.  They are not necessary, and bring *massive* costs and complexity to the table.

**Comment by BlackGad** [comment]
So no feature - no issues. Let's kill this feature by artificially limit its capabilities. And then be upset that feature is unpopular. 

You deeply mix build process and simultaneous IDE reaction because why not - same code. And now have performance issues because of huge codebases inside IDE. Second attempt to optimize this was filtering pipelines. Nice try! But slow intellisense reaction again beat you.

Why do you think that we do not realize your issues? 

Give us built in way to generate stuff. Source code it is not only C# files. But you insist that they are. And it feals like you fooling us, that's why so mad reaction on your messages. 

Anyway sorry for so emotional response :) From my point of view feature is awesome and I do not want to see it's failure.

**Comment by CyrusNajmabadi** [comment]
> Let's kill this feature by artificially limit its capabilities. 

The limitations are not artificial.  They're driven by a need to ensure that we can efficiently and effectively allow code to both introspect and inject itself into a running compilation.

> You deeply mix build process and simultaneous IDE reaction because why not - same code.

Yes.  Because that's what an IDE is, a system that provides "integrated development" experiences by leveraging things like compilers to understand code and provide features on it.  That's literally been how these tools have operated for 30 years.  SGs are the new thing here, not integrated, real-time analysis.  

> Give us built in way to generate stuff. 

There are built in ways to generate stuff.  Defining a process and running a pre or post build step is something msbuild has supported since *inception*.  People have been doing it for literally decades.   it's the standard normal way of doing things, that should be the default *unless* you have a domain need that involves compiler injection that no other system can solve for you.

> Anyway sorry for so emotional response

I don't know what to tell you.  Bringing emotions into this doesn't help.  We need to practically solve these areas by having people use the right tools and right options and the right layers.  There are very real and significant consequences for loading inappropriate tasks into the system in an inappropriate location.  

>that's why so mad reaction on your messages.

I'm sorry you're mad.  But your desires don't get to unilaterally trump the desires of other users.  For example, the ones that are extremely mad about things like performance.  We have to consider their needs as well, and draw safeguards around systems to prevent things from getting even worse and making even more of the ecosystem unhappy.

**Comment by geeknoid** [comment]
@CyrusNajmabadi You describe the cost of using a code generator to emit this state and recommend external tooling instead. However, looking at the overall developer experience (rather than just Roslyn execution time), I believe that in general the code generator is considerably cheaper in wall-clock time than a separate tool.

Running a separate tool involves starting processes and coordinating with their execution. And if the tool needs to understand the semantics of the source code, then the tool will be cranking up Roslyn internally to process the source code. This second use of Roslyn will be considerably more expensive E2E then would have been running the source generator in the first place as part of normal compilation activities.

Something we do in our code generator that emits non-C# files is to not run the generator in design-time environments. So as the user is typing in VS, we're not constantly spitting out the extra files. It would be more robust if this behavior was handled by the generator infra rather than our ad hoc solution, of course.

**Comment by filipnavara** [comment]
> I am curious about this use case: let's say you have a collection of .proto files from protobuffer, and you want to create a source generator that generates the .cs bindings. As far as I understand, this is one of the use cases intended for source generators (generating serialization bindings)

The .proto files are not .cs source files, so this is not necessarily a good fit for source generator. Of course it can be (ab)used for that but it doesn't involve the circular dependency where you need to parse the code first and then generate something based on the parsed code.

In contrast, for things like the JSON or XML serializer you often start with a C# class that is annotated with attributes. Thus you need to parse code first and then generated more code based on the parsed code, and then feed this new code back into the compiler.

> I mean, a while ago, when source generators where introduced, [they were advertised almost as the next .Net revolution](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/), and naturally, developers have began looking for new uses for source generators. Alas, that's not really the case and thus, the backslash.
>
> I think microsoft needs to clarify the intended use case of source generators, and specially, which use cases are not intended for, and why.

Agreed, we also abuse the source generators because we use them for something that could have been a build task. However, unlike a build task, they naturally produce output for design-time builds. While that's also possible with build task it's no longer a 5-line solution.

**Comment by ChrML** [comment]
I currently use Source Generators to generate Javascript files based on C# code. However the way I do it is quite hacky, due to to the current limitations:
- I create the JS file as a string and output C# code that embeds that string into the DLL as a static property.
- I have a very simple external tool that opens the DLL after build, finds the property and extracts the string into JS files
- Creating the JS files is quite expensive, so I added a condition in the csproj file to only run the source generator upon build (not when typing)

It works quite well, but I would like if this was the intended use case of source generators.

Some advantages I noticed with this approach:
- Very little build overhead, since the C# code is already parsed and the Roslyn syntax tree is already built.
- Very little tooling / bloat since source generators are already integrated with the build pipeline
- No additional steps required in the CI/CD pipeline

I believe the following changes would be needed to support this:
- Allow other resource files to be added by source generators
- Allow a way to decide if your source generator will run always on-change, or just on build (no need to generate resources like CSS continiously)
- Perhaps a consistent way to order the source generator execution

Based on the discussion above so far, it seems a bit like someone has decided "this is not what we designed the source generators to do, so it's not something we want to add". Which is a bit pity, because it really feels like source generators are 90% capable of this right now, and it just needs a bit of tweaking to get there.

The use cases for me are quite clear and useful, two smaller examples:
-  Swagger (OpenAI) definitions. Today Swagger uses runtime reflection and the XML docs output to generate your API documentation at the webserver at runtime. It could analyze your code and comments instead, output the JSON as static files as well as the index.html file for viewing it. Resulting in a smaller build and no runtime overhead. As a bonus, Swagger would be NativeAOT compatible when time comes for that.
- Generating TS definitions based on your API controllers for the front-end to use. Which could then be packed and distributed as a NuGet / NPM.

As an alternate approach, one could decide to not use Source Generator's for this, but come up with another name such as Custom Build Step or similar. However I see a lot of similarity, basically a Custom Build Step would be a Source Generator that you order, and disable during IDE use. The idea is that it anyway should have access to the same syntax tree as the C# compiler, to avoid having to parse the code multiple times during a build, and that it should be easy to add to your project from a NuGet package without fiddling with the MSBuild tasks.

**Comment by CyrusNajmabadi** [comment]
> While that's also possible with build task it's no longer a 5-line solution.

Seems like the efforts would be better spent on actually producing a 1-line solution there.  :)

I get the sense that other teams aren't necessarily interested in this space, so people have latched onto roslyn as being potentially a *workaround*.  But that's def not the goal of this feature space.  Something conceptually that can run before us (or after us) *should* be run that way.  It's much saner, cleaner, and **cheaper**.  Only things that need to literally run in-situ with the compiler should be in a generator.  

Now, unfortunately, i have no power to get other teams to care about those pre/post steps.  But my hope is that with enough community interest from people like yourselves, that could change.

**Comment by jaredpar** [comment]
> Definitely not. That's one of the most quintessential examples of what i would use another tool for

Disagree strongly. This is definitely an intended use case of generators. The C# team uses a generator to translate xml files into our syntax trees. That is very analogous to what is being described with .proto files here. 

> After all, most systems have no mechanism to inject code directly into a running compilation. And yet, generating serialization bindings is something totally feasible. That's been a solved problem for decades :)

By providing access to the `Compilation`, and all the associated APIs, the Roslyn system allows for significantly higher fidelity serializers to exist. Right or wrong in order to properly inspect C# code you need access to symbolic information and the most reliable way to get that is from the compilation.

**Comment by kg** [comment]
> > While that's also possible with build task it's no longer a 5-line solution.
> 
> Seems like the efforts would be better spent on actually producing a 1-line solution there. :)
> 
> I get the sense that other teams aren't necessarily interested in this space, so people have latched onto roslyn as being potentially a _workaround_. But that's def not the goal of this feature space. Something conceptually that can run before us (or after us) _should_ be run that way. It's much saner, cleaner, and **cheaper**. Only things that need to literally run in-situ with the compiler should be in a generator.
> 
> Now, unfortunately, i have no power to get other teams to care about those pre/post steps. But my hope is that with enough community interest from people like yourselves, that could change.

In our case at least, we're already running a source generator (and need to), and it's a question of whether we need an additional build step that parses everything again to generate some satellite data paired with the .cs. The ideal would be to just generate it all in one go, since we're already doing a bunch of the work inside of roslyn in our source generator.

We could do something like what @ChrML described, embedding the non-C# information inside of our generated C# as string literals or something, and then extract it. But that poses its own problems especially since you probably can't do that with pure MSBuild so now you're defining additional custom msbuild tasks in addition to all the other junk you have to write.

**Comment by tannergooding** [comment]
> Definitely not. That's one of the most quintessential examples of what i would use another tool for. And, indeed, that's how .proto works everywhere. After all, most systems have no mechanism to inject code directly into a running compilation. And yet, generating serialization bindings is something totally feasible. That's been a solved problem for decades :)

It's been "solved" in that solutions exist; but not solved in that there isn't "better" or "easier" solutions.

Generating bindings (serialization, interop, etc) are something that typically requires a compiler (at least a basic lexer, often something that can generate a proper AST); for 1 or both of the languages involved. 

I think the differentiator really comes down to how frequently the tool needs to run and to what degree it needs to interact with the C# compilation.

---------------------

In the case where you have `C#` and need more `C#`; source generators are obvious.

In the case where you have `C#` and need `Other`, it depends on what exactly you're doing. A lot of very simple generators may be better as some external tool that runs posterior to `csc` and which handles the IL instead. However, there are some complications there and can often be data that is lost in translation given that IL isn't 1-to-1 with C# and that many C# constructs may expand into non-trivial IL. A good example of this would be something that doesn't need to just process the public API surface, but also the method implementation details; such as a `C# -> HLSL` transpiler. Processing the IL is much more complex than processing the C# AST and in the worst case may be more akin to writing your own mini `ILSpy` to handle the various transforms. While it doesn't impact C# IntelliSense, it does impact the total end to end build and needs to run every time C# produces a new binary output.

In the case where you have `Other` and need `C#`, then I think it also depends on what you're doing. Take for example, https://github.com/dotnet/clangsharp which takes C/C++ headers and generates C# bindings. The easiest thing for this is to be an external tool that runs prior to `csc` because it just needs to generate correct C# code and doesn't need to actually parse or traverse it at all. It also does a ton of work and doesn't need to run frequently. Really it runs at the cadence of new releases of the underlying native library and so you're talking about a ton of work that is needed a couple times a year. Another tool, however, may need to run more frequently and may have more direct impact on IntelliSense or other important features.

So I think we end up having the need for `source generators` to work for `C# -> C#`, `Other -> C#`, and `C# -> Other` (e.g. this issue). We then need to differentiate as to the cadence the feature needs to run. This could be `every keystroke`, it could be `every time a new assembly is being produced`, it could be `every time a particular file is changed`, it could be `ad-hoc once a week/month/year`. The first two both seem like cases where `source generators` are a very viable candidate and where having the right integration points to ensure they aren't running unnecessarily is "ideal". The third seems like a potential `source generator` candidate but also seems like maybe something that would be good to have some tooling considerations around. Perhaps such generators are only passed to `csc` by `msbuild` when an up to date check fails and some NuGet feature could help make this work implicitly. The fourth is very similar but also is the type where it runs so rarely that it probably doesn't matter too much and simply depending on `Microsoft.CodeAnalysis` for a standalone tool would be fine.

**Comment by ChrML** [comment]
> > > While that's also possible with build task it's no longer a 5-line solution.
> > 
> > 
> > Seems like the efforts would be better spent on actually producing a 1-line solution there. :)
> > I get the sense that other teams aren't necessarily interested in this space, so people have latched onto roslyn as being potentially a _workaround_. But that's def not the goal of this feature space. Something conceptually that can run before us (or after us) _should_ be run that way. It's much saner, cleaner, and **cheaper**. Only things that need to literally run in-situ with the compiler should be in a generator.
> > Now, unfortunately, i have no power to get other teams to care about those pre/post steps. But my hope is that with enough community interest from people like yourselves, that could change.
> 
> In our case at least, we're already running a source generator (and need to), and it's a question of whether we need an additional build step that parses everything again to generate some satellite data paired with the .cs. The ideal would be to just generate it all in one go, since we're already doing a bunch of the work inside of roslyn in our source generator.
> 
> We could do something like what @ChrML described, embedding the non-C# information inside of our generated C# as string literals or something, and then extract it. But that poses its own problems especially since you probably can't do that with pure MSBuild so now you're defining additional custom msbuild tasks in addition to all the other junk you have to write.

I agree. I believe access to the `Compilation` object and all the semantics that the C# compiler uses to compile the application is one of the main reasons we use the source generator this atm. There's still the MSBuild task to extract the data, unfortunately.

For my use case, I would be happy with any other solution that: Runs on build, has access to `Compilation` + semantics, can be added to a user's build tasks just by installing a NuGet (adding one PackageReference to csproj) and that can output resource files that will be included with the build. 

Unfortunately, I don't think such solution exists at the moment. The source generator was the "close enough" solution that I found. But I'm happy if someone has a better idea :)

**Comment by ChrML** [comment]
> I guess that generating external files, like typescript or C files could be offloaded from the compiler main pipeline, (which is on tight performance budget). The parsed C# tree could be available as read only clone on separate thread. Perhaps some new event to subscribe to ?

That would work for me.

**Comment by dotMorten** [comment]
I use the CsWin32 code generator quite a lot. It's a pure joy to work with and is the "other => C#" kind. I often find myself needing another Win32 method to call, and I just go and edit the `NativeMethods.txt` file and instantly the new method is available for me to call and intellisense ready to help me figure out how. It's an amazing developer experience, even if that's not the "proper" way to use generators. And in this case, this generator only needs to run when the input file changes, which is probably the common case for "other => C#" scenario.
I'd be curious for a solution that would allow the CSWin32 generator to follow "proper" patterns, since it sounds like @CyrusNajmabadi says it is not what generators are meant for and is currently an abuse of the system.

I do find that "C# => Other" and "Other => Other" is quite useful as well (As an example a list of image files gets converted into a C# enum, as well as some XAML static resources with the image sources pre-defined).

**Comment by jaredpar** [comment]
> I'd be curious for a solution that would allow the CSWin32 generator to follow "proper" patterns, since it sounds like @CyrusNajmabadi says it is not what generators are meant for and is currently an abuse of the system.

It is not an abuse of the system. It is part of the intended purpose.

**Comment by CyrusNajmabadi** [comment]
> Disagree strongly. This is definitely an intended use case of generators. The C# team uses a generator to translate xml files into our syntax trees. That is very analogous to what is being described with .proto files here.

This is actually a very good case where we run into problems.  Much of my perf work (including the last week in PRs that are open today) is just dealing with the fallout of that.  For example, how poorly the decl table behaves here, and how challenging this is to thread through the higher stack layers.  

I would *far* prefer this not be a generator**.

--

** Caveat:  I find this valuable as a generator *solely* so we *dogfood* the product.  Having this generator shows us all the problems that generators cause in our system, and forces us to feel that pain firsthand (including impact on features).  In that regard, i think it's good to have it.  It also prevents us from introducing any optimizations (like: "roslyn has no generators, so just disable all those paths") that are not customer-representative, which coudl lead us into thinking things are good.

However, from a design perspective, i *def* consider the syntax-generator space and others like that we might do in the future (like the bound-nodes, etc.) to be something that actually strongly *demonstrates* why generators are the wrong choice here.

**Comment by CyrusNajmabadi** [comment]
> I just go and edit the NativeMethods.txt file and instantly the new method is available for me to call and intellisense ready to help me figure out how. It's an amazing developer experience

I strongly *want* "amazing developer experience"(s).  Don't take my position to mean that that's not a goal.  My general point though is solving problems at the right layers with the right tools.  Those layers include, but are not limited to, "language, compiler, pre-tools, post-tools, runtimes, jits, aot, etc.".  I like .net because it can do great things, figuring out good places in this stack for user-facing features to live.  I also get that once there is a hammer, lots of things can look like nails :)   My point here is that if we want things like pre-tools or post-tools, then we need better *systems* for that.  Systems that can provide the bulk of the value, likely with safeguards and guardrails, to prevent the bulk of the problems.  I'm a strong proponent of narrow-path solutions that can solve the 95% case, but at orders of magnitude less cost than trying to reach 98-99-100%.

**Comment by stephentoub** [comment]
@CyrusNajmabadi, I read your comments as being about your opinions and not necessarily representative of any larger group of which you're a part, is that correct?  The way the comments have been written suggests they're the general stance of the C# language team, the C# compiler team, the .NET team in general, etc., and I want to clarify for anyone reading here that that is not the case.

**Comment by CyrusNajmabadi** [comment]
> That would work for me.

Agreed.  Analyzers, for example, have a model which is much more suitable here.  They are an *out of band* eventing system, which do not contribute to a separate compilation step, but which only get activated if something like 'diagnostics' actually needs them.  This sort of approach is *much* better than the generator approach which sits *underneath* the compiler, and thus has costs always paid by that compiler.

**Comment by CyrusNajmabadi** [comment]
@stephentoub I thought me views aligned with the teams as of the last meetings we had on the topic of generating non-source artifacts.  It does sound like those views may have changed.   My personal views (and those about the IDE space) remain the same.   We'll likely need to have meetings on this to actually flesh out how this could work, and be done in a way that doesn't compound issues.  :)

**Comment by CyrusNajmabadi** [comment]
> Perhaps such generators are only passed to csc by msbuild

Yes.  I'm *strongly* in favor of a system that codifies these are explicit pre/post compile steps, with then very clear semantics on who runs, when they run, and what they can generate.  I would even go as far as to potentially have those tools block anyone (mis)using this feature for `Other => Other` generation (which i'm guessing will start happening).  

For these new cases, your tool must either be:
1. generating C# for that compilation.
2. taking the C# and producing something outside the compilation.


I like msbuild integration here because now your inputs/outputs need to be clearly declared to properly fit into things like determinism through these layers.

**Comment by CyrusNajmabadi** [comment]
> I do find that "C# => Other" and "Other => Other" is quite useful as well (As an example a list of image files gets converted into a C# enum, as well as some XAML static resources with the image sources pre-defined).

I am reallllly wary about `Other => Other`.  csc now becomes a general extension host for anyone wanting to transform arbitrary files?  That seems quite problematic to me, and makes me strongly reconsider if we shoudl even call this csc versus ".net tool runner host" :D

In your case ("a list of image files gets converted into a C# enum, as well as some XAML static resources") that seems morally like exactly what a pre-step would be.  That step could utilize the roslyn apis of course (to generate syntax), but doesn't seem to have anything intrinsic to it that means it should run inside the compilation (it doesn't seem like it needs to understand the users own source code meaning).  If i'm misunderstood something, please let me know.

**Comment by CyrusNajmabadi** [comment]
@dotMorten That def does make things feel better.  :)  It would at least feel like something i could explain to users and still feel good about a block on.  (if i'm being unclear: your use case would work.  teh general `only other => only other` is something i would prefer be blocked)

**Comment by tannergooding** [comment]
> Yes. I'm strongly in favor of a system that codifies these are explicit pre/post compile steps

Noting that in the world I was describing here many of these may still be "Roslyn Source Generators". The difference is that the build tooling helps ensure that they don't run unnecessarily.

> You def don't need a lexer or parser to generate C#, esp. for a case like proto generation.

Right, and its often simpler to generate C# using simple string formatting than using creating syntax trees with trivia/etc. However, there are cases where C# is the source language and where having the AST is the best thing (e.g. going from `C# -> HLSL`). The difference is that it doesn't impact C# IntelliSense and so doesn't need to run every keystroke but it does need to run each time an assembly is produced. The overhead of running separate post tools that depend on `Microsoft.CodeAnalysis` and ensuring that the handling of all inputs match what was given to `csc` can be quite non-trivial and is easily broken over time as the `Csc` build task is versioned so new tools defaulting to a source generator that only runs in such compilation producing scenarios would be the preferred/obvious choice for many users.

**Comment by CyrusNajmabadi** [comment]
> Right, and its often simpler to generate C# using simple string formatting than using creating syntax trees with trivia/etc. However, there are cases where C# is the source language and where having the AST is the best thing

Absolutely.  My point is that that comes from using the Roslyn APIs, not necessarily being *in* a Roslyn compilation :)

LSIF is a prime example of this.  We need syntax/semantics/operations/etc.  But we just get that through teh normal roslyn API surface area, without that artifact having to be generated from *Within* the compilation itself.

>and ensuring that the handling of all inputs match what was given to csc

Yes.  This seems like a very unfortunate issue with the systems *today*.  I would love if such a concept were made first class, so that one could run perfectly without issues in that space.

> The overhead of running separate post tools

Note: conceptually (though it woudl be more work) i don't think there needs to be overhead here *if* compiler-server is involved (but it would be if only directly using a fresh instance of csc that didn't talk to that).  Compiler-server could certainly reuse compilation instances across the tools if needed.

**Comment by lsoft** [comment]
related point:

many `C# -> Other` SG can be converted into a custom `exe` (to be a part of build process), if people will have an inter process access from their `exe` to `Compilation` object (which is already exists in Roslyn process in VS).

such IPC-Compilation concept would not completely solve the subject (at least not my own need), but it would mitigate its poignancy.

**Comment by vpenades** [comment]
@CyrusNajmabadi regarding my original use case of .proto to c# ...  the proto files are basically structure definitions, that are reflected to PoCo objects in c# ... so it makes sense that you add a new field in the proto file, and it becomes available in the c# code.

But generating the c# code from .proto implies executing an external command line tool, so I suspect, given the considerations you've explained about how source generators work, _it's probably not the best way of doing things_.

Now, I would be happy to have some kind of behavior like T4 templates: that is, the code generations only happens when I have finished editing the t4 file and I switch tabs.  In the case of .proto files, you don't need access to the AST.

As you have stated, many use cases can be solved by pre/post solutions, which generally involve crafting arcane msbuild spells...  so my suggestion would be, if such a thing is possible, to introduce new interfaces for pre and post use cases which would _divert_ developers from use the c#->c# source generators

**Comment by riverar** [comment]
I don't have a horse in this race but just wanted to say I too was unpersuaded by the `C# > Other` and `Other > Other` scenarios presented above. I suspect folks are really just looking for an alternative to working with MSBuild and its Tasks for build workflow orchestration. I can sympathize, but I'm not sure increasing Source Generator scope (and complexity) to support those scenarios is what the ecosystem needs.

**Comment by DomenPigeon** [comment]
@CyrusNajmabadi just wanted to thank you for advocating for the use of MSBuild, because of your comments I was forced to go check out how MSBuild actually works and what capabilities it has and I was amazed about how little I actually knew about it and how strong of a tool it is.

For the case I have mentined above, now that I know what is MSBuild capable of, I don't need the source generators for it, because I can generate all the mirror JS files from C# via reflection.

The one thing that would still be hard, as far as I now understand the MSBuild, (as was mentioned above in some comment), is to generate files (C# or other) based on the syntax tree or semantic model, which would be hard to get inside a task but is easily accesable when writting source generators.

**Comment by CyrusNajmabadi** [comment]
> The one thing that would still be hard, as far as I now understand the MSBuild, (as was mentioned above in some comment), is to generate files (C# or other) based on the syntax tree or semantic model, which would be hard to get inside a task but is easily accesable when writting source generators.

Correct. This is the area where SGs excel (and do things that would be extraordinarily difficult to do from outside the compiler). :-)

**Comment by Perksey** [comment]
Encountered a need for this today to wield Blazor's WASM tooling to generate a native P/Invoke trampoline to account for scenarios that the current P/Invoke tooling in Blazor doesn't handle very well.

However, I fully appreciate that analogues to this feature in other languages (such as the Rust proc_macro) also do not implement this because fundamentally it doesn't make sense for a A->B compiler to also generate C to then be fed into another compiler, they're fundamentally different stages of the pipeline.

Perhaps a happy medium could be allowing the generation of a metadata file from a source generator to allow SGs to provide information about their execution retrospectively, to then be picked up by another tool, using an opt-in parameter on csc.

**Comment by BlackGad** [comment]
You can generate cs file with block comment on full json/xml content) make extension something like foo.metadata.cs. on pre build step (but after generation) rename file to *.json/*.xml with comment block cut. 

Of course it is a trick but still if you really need this - would work.

**Comment by mhmd-azeez** [comment]
@Perksey I have been able to successfuly do that by writing a custom MSBuild task:
https://github.com/extism/csharp-pdk/pull/4

**Comment by Kukkimonsuta** [comment]
Would allowing generating non-source files only within `RegisterImplementationSourceOutput` allow this feature to go through? Because I think most/all use cases mentioned here would be fine with that.

Alternatively if we really must avoid writing files at all cost during this process, could there be a mechanism to push information back to msbuild so a msbuild task can pick it up and further process however/whenever needed? That would allow code inspection during compilation while allowing generating whatever non-code files based on that after compilation.

**Comment by jaredpar** [comment]
> Would allowing generating non-source files only within RegisterImplementationSourceOutput allow this feature to go through?

No. This feature still has a number of open issues and likely requires changes to other products for us to ship. For example non-source file outputs must be written to disk. That represents dynamic build outuputs from the compilation task. That is not supported by MSBuild today and would lead to basic build correctness issues. Dealing with this is an active discussion topic between us and the MSBuild team. 

This is a feature we're very much looking at for .NET 9 but there is not a simple fix.

**Comment by jaredpar** [comment]
Want to elaborate on the problem this presents for MSBuild. The purpose of this feature is to allow source generators to produce non-source files outputs. These outputs will be written to disk for later phases of the build to consume. 

MSBuild tracks both inputs and outputs to targets. At a high level when evaluating a given target for a project MSBuild will determine if the inputs to the target are the same as the last evaluation and the outputs still exist, if so then it can skip the target entirely. This is how incremental builds, fast up to date, etc ... function in MSBuild. This means though that if you do a build of `console.csproj`, then say delete `obj\net472\console.dll` a rebuild will re-run the compiler because MSBuild can see the output file is missing hence the target is out of date.

MSBuild support though requires that all of the inputs and outputs are known at the time the target is evaluated. That is where the check occurs. The issue with non-source file outputs is they are determined dynamically by the generators based on the content of the `Compilation` they are given. That does not fit into MSBuild's existing tracking model. These would all untracked outputs and features like rebuild, FUTD, etc ... would be broken. 
 
I understand that some contributors here may see that as an acceptable trade off but we do not. Really hard to push through a compiler feature that is fundamentally incompatible with our basic build principles. It's a recipe for customer confusion and lots of feedback issues. 

There are a couple of ways this could be fixed: 

1. Break up `CoreCompile` target into a series of targets: `CoreCompileGetOutputs` and `CoreCompileCompile`. The `CoreCompileGetOutput` targets runs the generators enough to get the output list and provide that to `CoreCompileCompile` which runs the compiler. The issue with this approach is that `CoreCompileGetOutputs` must always run. It must be constructed in a way that it always fails the up to date check. That means generators run, even in a limited form, on rebuild scenarios. Not ideal. 
2. Change MSBuild such that it supports targets with dynamic outputs, and likely dynamic inputs. The `CoreCompile` target could be changed to take advantage of this and non-source file generators would fit naturally into that scenario.

At the moment (2) is the preferred approach and we're digging into its viability.

MSBuild tracking issue: https://github.com/dotnet/msbuild/issues/701

**Comment by rainersigwald** [comment]
> MSBuild support though requires that all of the inputs and outputs are known at the time the target is evaluated.

Nit: it's at _execution_ time, **just before** executing the tasks inside the target (step 6 of [the target order determination](https://learn.microsoft.com/visualstudio/msbuild/target-build-order?view=vs-2022#determine-the-target-build-order).

(It could have been at evaluation time but it's slightly more dynamic than that--if it was at evaluation, plan 1 wouldn't work.)

**Comment by Perksey** [comment]
This does feel like it may invoke a future feature request for `RegisterIncrementalImplementationSourceOutput` or something equivalent, as I don't think everyone will want to regenerate every time, but this sounds on the right track.

**Comment by riverar** [comment]
@jaredpar I remember being unconvinced that this feature was truly needed (in MSBuild). Did you or the team come up with scenarios that are driving this current work?

**Comment by kg** [comment]
For browser scenarios, CSP rejects attempts to dynamically generate code or run code that wasn't loaded from a trusted network source, so if you are (for example) generating C#/JS interop bindings, it is ideal if you can produce a paired .cs file and .js file from the same inputs in one go. Embedding the JS in a string literal would violate CSP, so the only alternatives are to have a separate tool that generates the .js files from the source metadata, or to embed the JS in the generated C# somehow and extract it with an msbuild task.

IIRC we ended up working around the lack of good solutions by using an interpreter instead of generating JS, even though it's slower. I don't know if we would actually rewrite our code to use this feature when it arrives (we were told to expect it a long time ago).

I also saw someone suggest an analyzer+source generator as a way to implement https://github.com/dotnet/csharplang/discussions/7529 which would require the generator be able to generate string tables as secondary outputs. (I don't think that is actually a solution for the use case, but it was suggested there.)

**Comment by jaredpar** [comment]
@riverar 

>  I remember being unconvinced that this feature was truly needed (in MSBuild). 

Can't tell which feature you're speaking about here: non-source outputs in compiler or dynamic output tracking in MSBuild. 

If it's non-source outputs, it's still not 100% clear if it's fully needed. There are many scenarios where it would be a strong benefit to the ecosystem. There are ways to work around this by just writing more MSBulid code but it would be a lot better integrated in the source generator pipeline. Essentially there is enough benefit here for us to be doing deeper investigation. 

If it's msbuild tracking dynamic output,, I outlined one of the more important scenarios above. Effectively a general build principal for us is deleting build output should cause rebuild to re-execute compilation. Lacking MSBuild features or the targets trick I mentioned this can't happen. That's a non-starter for us right now.

**Comment by eerhardt** [comment]
I just wanted add a +1 to this scenario and link to the .NET Aspire Components use case for this: https://github.com/dotnet/aspire/issues/1146.

.NET Aspire Components allow app settings to be configured in `appsettings.json` files. We added a feature to Visual Studio's JSON editor to support augmenting the JSON intellisense with JSON schema segments coming from referenced NuGet packages. For example, when a project references the [Aspire.StackExchange.Redis](https://www.nuget.org/packages/Aspire.StackExchange.Redis) package, and the dev opens the `appsettings.json` file, they get auto-completion and intellisense for the Redis options in VS (note we want to add this feature to VS Code as well):

![image](https://github.com/dotnet/roslyn/assets/8291187/ab561f40-bee7-4b37-8f6f-69d50d2f6a04)

The input to these JSON schema files is .NET Types, in this case the class is the StackExchange.Redis [ConfigurationOptions](https://github.com/StackExchange/StackExchange.Redis/blob/510b5e21b256b5d4226c79062b8ecfc4d2265817/src/StackExchange.Redis/ConfigurationOptions.cs#L210-L226) class. As we update this class, I'd want the JSON schema files to be automatically updated, instead of having to hand-craft the files that we do today. This seems like a perfect fit for a Roslyn Source Generator, if it supported outputting non-source files.

**Comment by CyrusNajmabadi** [comment]
@eerhardt why does this need to be a SG thought?  Nothing in the same app will consume that same json schema right?  So this could just be a trivial tool that runs over the compilation as a *post-step* and produce the json schema file.

**Comment by BlackGad** [comment]
> @eerhardt why does this need to be a SG thought?  Nothing in the same app will consume that same json schema right?  So this could just be a trivial tool that runs over the compilation as a *post-step* and produce the json schema file.
> 
> 

Because of double syntax tree build maybe?) With proper cache SG could do this automatically. Also it will be convenient to encapsulate this logic to separate nugget package so proj file will be untouched

**Comment by CyrusNajmabadi** [comment]
> Because of double syntax tree build maybe?)

1. with a compiler server, this should be a non-issue.
2. this motivates a *different* technology that is nto *source-generators*.  This would be something like a post-generator tech.  i'm much more ok with that.  

Note: i'm fairly sure people want a post-generator, since SGs can't see the results of other SGs.  So if you wrote an SG here, for example, your json-schema could not include any information about things generated by other generators, which is not what i think people want.

**Comment by eerhardt** [comment]
> @eerhardt why does this need to be a SG thought? Nothing in the same app will consume that same json schema right? So this could just be a trivial tool that runs over the compilation as a _post-step_ and produce the json schema file.

It doesn't **need** to be a SG. To unblock the scenario my plan is exactly as you suggest - a post-compilation step. It just feels more natural to do this during `csc` and using Roslyn Source Generators than it does to make custom MSBuild tasks/targets/etc.

**Comment by Khitiara** [comment]
definitely seems like a general way to hook into the compiler to access syntax and the semantic model during other tasks/targets is what people actually need. can say that would handily cover my use-cases for this feature too, as long as that can be made to run early enough to create embedded resources

**Comment by CyrusNajmabadi** [comment]
> definitely seems like a general way to hook into the compiler to access syntax and the semantic model during other tasks/targets is what people actually need. 

Yeah, this is definitely possible, though we coudl likely make things better here by making the tasks easier to write.  As an example, the roslyn IDE does this ourselves in order to generate LSIF files from the built compilations.  If we could extract out the portion of that taht handles getting the same data and then operating on that, that would likely be the most valuable thing here.

IMO, given a compielr server, you have the best of all worlds:

1. the server keeps the data alive (compilations, references, trees) so you're not paying the expense multiple times.
2. you run after all real source generation is done.  so you see the *complete* view of hte code.  not the primordial view.
3. you are just a normal msbuild task with proper inputs/outputs.  
4. you incur no costs deep in the compiler.
5. you can use *any* generation mechanism.  you're not dependent on things like ForAttributeWithMetadataName.

**Comment by eerhardt** [comment]
> So this could just be a trivial tool that runs over the compilation as a post-step and produce the json schema file.

I took a stab at writing this today, and I don't think this is "trivial". How exactly would you do this? What I tried:

* Make an MSBuild Target/Task that hooks into `TargetsTriggeredByCompilation`
* Reference `Microsoft.CodeAnalysis.*` assemblies in the Task
* Pass `@(ReferencePaths)` and `@(IntermediateAssembly)` to the Task

The issue I immediately hit is that the `Microsoft.CodeAnalysis.*` assemblies are not loaded by MSBuild. So I need to find them somehow myself, either bring my own or have MSBuild pass the location and hook AssemblyLoadContext [like Arcade does](https://github.com/dotnet/arcade/blob/29736c3a40dc33e94b8a7fdf72dad6d441f9e46f/src/Microsoft.DotNet.GenFacades/RoslynBuildTask.cs#L23-L24).

Is using a `Task` wrong here and instead should be making a `.exe` that does this instead? I would still have the problem of how to load the `Microsoft.CodeAnalysis` assemblies.

> with a compiler server

How do you make a custom task or tool that uses a compiler server? Are there examples?

**Comment by jaredpar** [comment]
> I took a stab at writing this today, and I don't think this is trivial. How exactly would you do this? 

I don't think it's realistic to do. 

> How do you make a custom task or tool that uses a compiler server? Are there examples?

You can't really. The only task that can access the compiler server is the `Csc / Vbc` tasks. That is unlikely to change. 

I do not think hosting the compiler yourself is a realistic path forward. We've seen **many** teams attempt to do this over the years and it does not end well for them.

**Comment by lsoft** [comment]
This article contains many use cases. May be it is enough now to generalize... I don't want to seem annoying, but let me ask please: is there any plan in Net9 dev cycle to work about this topic?

if no, what is about some kind of workaround: many of us will be happy if we will have a way to touch `Compulation` (`SemanticModel`, `Document`, `SyntaxNode` and other Roslyn stuff) **efficiently** somewhere in the build cycle (prebuild?). (https://github.com/dotnet/roslyn/issues/57608#issuecomment-1543333499)

**Comment by CyrusNajmabadi** [comment]
@Misiu When there are updates, we'll def upate the issue.  Thanks :)

**Comment by fr4gles** [comment]
Would love to hear any update on this task in 2026 🙂
